### PR TITLE
Fix 22831 - Check signature of extern(C) main functions

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Enforce that the `main` function uses (most likely) valid arguments / return types. The spec / C standard denotes the following signatures:

```d
int main() { ... }
int main(int, char**) { ... }
```

The implemented checks are more lenient to accomodate for common deviations from the standard. See the DDOC comment of `checkMain()` for a list of accepted extensions. Exotic platforms that expect a different signature can circumvent the checks using  `pragma(mangle, "main")`.

See e.g. https://stackoverflow.com/questions/2108192/what-are-the-valid-signatures-for-cs-main-function

CC @pbackus

### CODE REVIEW

**Comment by dkorpel** [comment]
Is this ready for review?

**Comment by MoonlightSentinel** [comment]
Yes

File: src/dmd/func.d [hunk 1]
```diff
@@ -2609,35 +2609,109 @@ extern (C++) class FuncDeclaration : Declaration
...
+            assert(this.inferRetType);
+            return;
+        }
+
+        /// Checks whether `t` is equivalent to `char**`
+        /// Ignores qualifiers and treats enums according to their base type
+        static bool isCharPtrPtr(Type t)
+        {
+            auto tp = t.toBasetype().isTypePointer();
+            if (!tp) return false;
```
Comment (dkorpel) [inline src/dmd/func.d:2649]: `return` statement should be on its own line

File: src/dmd/func.d [hunk 2]
```diff
@@ -2609,35 +2609,109 @@ extern (C++) class FuncDeclaration : Declaration
...
+
+            tp = tp.next.toBasetype().isTypePointer();
+            if (!tp) return false;
```
Comment (dkorpel) [inline src/dmd/func.d:2652]: ditto

File: test/compilable/betterCarray.d [hunk 3]
```diff
@@ -4,7 +4,8 @@
 
 import core.stdc.stdio;
 
-extern (C) int main(char** argv, int argc) {
```
Comment (dkorpel) [inline test/compilable/betterCarray.d:7]: Why remove these parameters?

**Review by dkorpel — CHANGES_REQUESTED** [review]

File: test/compilable/betterCarray.d (hunk 3 above)
Comment (MoonlightSentinel) [reply test/compilable/betterCarray.d:7]: Because the signature was wrong and the parameters aren't used in the function body.

Comment (dkorpel) [reply test/compilable/betterCarray.d:7]: I saw the latter, but not the former.

File: src/dmd/func.d [hunk 4]
```diff
@@ -2609,35 +2609,109 @@ extern (C++) class FuncDeclaration : Declaration
...
+                    t.nextOf().ty != Tarray ||
+                    t.nextOf().nextOf().ty != Tchar ||
+                    fparam0.storageClass & (STC.out_ | STC.ref_ | STC.lazy_))
+                {
+                    argerr = true;
+                }
             }
+
+            if (tf.parameterList.varargs || nparams >= 2 || argerr)
+                error("parameters must be `main()` or `main(string[] args)`");
```
[the remark this answers is not in the document]
Comment (MoonlightSentinel) [reply src/dmd/func.d:2676]: Agreed, the current wording is not ideal. Maybe:

```D
error("parameter list must be empty or `(string[] args)`");
```
or
```D
error("must accept zero or one parameter of type `string[]`");
```

or

...?

[the remark this answers is not in the document]
Comment (dkorpel) [reply src/dmd/func.d:2676]: Yes, I think the first one is better, or the second one with a small change: "must accept zero *parameters* or one parameter of type `string[]`"

File: src/dmd/func.d [hunk 5]
```diff
@@ -2609,35 +2609,109 @@ extern (C++) class FuncDeclaration : Declaration
...
+     +
+     + The following signatures are allowed for a `D main`:
+     + - Either no or a single parameter of type `string[]`
+     + - Return type is either `void`, `int` or `noreturn`
+     +
+     + The following signatures are standard C:
+     + - `int main()`
+     + - `int main(int, char**)`
+     +
+     + This function accept the following non-standard extensions:
```
Comment (RazvanN7) [inline src/dmd/func.d:2623]:
```suggestion
     + This function accepts the following non-standard extensions:
```

**Review by RazvanN7 — APPROVED** [review]

**Review by dkorpel — APPROVED** [review]

**Comment by WalterBright** [comment]
This caused a regression:
https://issues.dlang.org/show_bug.cgi?id=22961

**Pull request merged**